File: easy/most_common_safe_word_test.go
package easy

import (
	"strings"
	"testing"
)

func TestMostCommonSafeWord(t *testing.T) {
	cases := []struct {
		paragraph string
		banned    []string
		expected  string
	}{
		{"Bob hit a ball, the hit BALL flew far after it was hit.", []string{"hit"}, "ball"},
		{"a.", []string{}, "a"},
		{"a b.b", []string{}, "b"},
		{"Bob. hIt, baLl", []string{"bob", "hit"}, "ball"},
		{"Apple apple APPLE!", []string{"apple"}, ""},
		{"Wow! wow? wow. really, really... REALLY!", []string{"wow"}, "really"},
		{"test test test test", []string{}, "test"},
		{"!@#$%^&*()", []string{}, ""},
		{strings.Repeat("word ", 1000), []string{}, "word"},
		{"apple banana apple banana cherry", []string{"banana"}, "apple"},
	}
	for _, c := range cases {
		t.Run("", func(t *testing.T) {
			result := MostCommonSafeWord(c.paragraph, c.banned)
			if result != c.expected {
				t.Errorf("MostCommonSafeWord(%q, %v) == %v, expected %v", c.paragraph, c.banned, result, c.expected)
			}

		})

	}
}

File: test/easy/most-common-safe-word.test.ts
import { describe, test, expect } from "bun:test";
import { mostCommonSafeWord } from "@/src/easy/most-common-safe-word";
describe("mostCommonSafeWord", () => {
    test.each([
        [
            "Bob hit a ball, the hit BALL flew far after it was hit.",
            ["hit"],
            "ball",
        ],
        ["a.", [], "a"],
        ["a b.b", [], "b"],
        ["Bob. hIt, baLl", ["bob", "hit"], "ball"],
        ["Apple apple pear APPLE!", ["apple"], "pear"],
        ["Wow! wow? wow. really, really... REALLY!", ["wow"], "really"],
        ["test test test test", [], "test"],
        ["apple banana apple banana cherry", ["banana"], "apple"],
    ])(
        "\n\tGiven: a paragraph %s.\n\tAnd: a list of banned words: %p.\n\tThen: the most common word is %s",
        (paragraphs: string, banned: string[], output: string) => {
            const result = mostCommonSafeWord(paragraphs, banned);
            expect(result).toBe(output);
        },
    );
});
